### ISSUE TITLE
Kitematic/Docker container list getting out of sync

### ISSUE THREAD

**Issue by alankent** [docker/kitematic#1349]
I am not sure the exact sequence causing the problem. But I have a container with a web application in it. The container exposes port 80 and the LiveReload port. It starts up fine. I then have to mount the volumes it has (to sync to the host OS file system). I pick the first volume and select "enable volume". 

This used to work fine, but now (I have changed the Dockerfile to add Grunt and the LiveReload port) when I add volume, it shuts down the container but fails to restart it saying 0.0.0.0:80 on the bridge is still in use. So I cannot restart the container.

After that, Kitematic goes a bit wonky. I cannot remove the container using the GUI. I cannot start it either. I remove the container using the command line, but Kitematic does not realize it is gone. So I cannot use Kitematic to launch a new container with the same name, and clicking on the cross to delete the container reports the 'id' is not in use.

So basically I cannot start the container up any more with a mounted volume using Kitematic. I am going to try and rewind my Dockerfile to an older version and see if it starts working again - to see if it is Grunt/LiveReload somehow causing the problem.

I find that every so often Kitematic seems to get its container list out of sync with Docker, and there is no "refresh" or similar to force it to check its list against the Docker engine to get the "real" list of containers. E.g. during builds, sometimes a stray container appears in Kitematic and wont go away. I cannot delete it because it says "id X does not exist".

**Comment by alankent** [comment]
I forgot to say next to the container id it says "Updating" in a blue box as well. Not sure what that means.

And I can confirm when I removed Grunt and LiveReload (removing "EXPOSE 35729" as well), I started being able to mount volumes again - the container restarted reliably. I then just added the EXPOSE command back in. I had the "docker run" going in a window, and I managed to capture this output on the console when it tried to shutdown the container. (I suspect this is MariaDB shutting down, but including in case useful.)

(Oh, note there is no running process listening on the port I exposed - "grunt watch" when run should start listening on that port, but I was going to start that up from the command line after the container is running.)

    Status information:

    Current dir: /var/lib/mysql/
    Running threads: 0  Stack size: 294912
    Current locks:
    lock: 0x7f79c0349088:

    lock: 0x7f79c0331f48:

    lock: 0x7f79c0348f08:

    lock: 0x7f79c0348d88:

    lock: 0x7f79c0348a88:

    lock: 0x7f79c0348908:

    lock: 0x7f79c0348788:

    lock: 0x7f79c0348608:

    lock: 0x7f79c0331408:

    lock: 0x7f79c0348488:

    lock: 0x7f79c0348308:

    lock: 0x7f79c0348188:

    lock: 0x7f79c5bf9e88:

    lock: 0x7f79c5bf9b88:

    lock: 0x7f79c5bf9a08:

    lock: 0x7f79c5bf9888:

    lock: 0x7f79c5bf9708:

    lock: 0x7f79c5bf9588:

    lock: 0x7f79c5bf9408:

    lock: 0x7f79c5bf9288:

    lock: 0x7f79c5bf9108:

    lock: 0x7f79c0331188:

    lock: 0x7f79c0330f08:

    lock: 0x7f79c5bf8c88:

    lock: 0x7f79c5bf8b08:

    lock: 0x7f79c5bf8988:

    lock: 0x7f79c0330b48:

    lock: 0x7f79c0330c88:

    lock: 0x7f79c5bf8808:

    lock: 0x7f79c0330a08:

    lock: 0x7f79c5bf8688:

    lock: 0x7f79c03308c8:

    lock: 0x7f79c0330788:

    lock: 0x7f79c5bf8508:

    lock: 0x7f79c0330648:

    lock: 0x7f79c0330508:

    lock: 0x7f79c03303c8:

    lock: 0x7f79c0330288:

    lock: 0x7f79c0330148:

    lock: 0x7f79c58e6ec8:

    lock: 0x7f79c58e6d88:

    lock: 0x7f79c58e6c48:

    lock: 0x7f79c58e6b08:

    lock: 0x7f79c58e6748:

    lock: 0x7f79c58e69c8:

    lock: 0x7f79c58e6888:

    lock: 0x7f79c58e64c8:

    lock: 0x7f79c58e6388:

    lock: 0x7f79c58e6248:

    lock: 0x7f79c58e6608:

    lock: 0x7f79c5a30f88:

    lock: 0x7f79e609f388:

    lock: 0x7f79e606f588:

    lock: 0x7f79e606e188:

    lock: 0x7f79e606c388:

    lock: 0x7f79e606af88:

    lock: 0x7f79bfc1cd88:

    lock: 0x7f79e60fd388:

    lock: 0x7f79e609c388:

    lock: 0x7f79e605e788:

    lock: 0x55a40aed4d80:

    lock: 0x55a40aed4c80:

    lock: 0x55a40aed3a80:

    lock: 0x55a40aed3980:

    lock: 0x55a40aed3880:

    lock: 0x55a40aed3c80:

    lock: 0x55a40aed1a80:

    lock: 0x55a40aed4b80:

    lock: 0x55a40aed2280:

    lock: 0x55a40aed2580:

    lock: 0x55a40aed2180:

    lock: 0x55a40aed2480:

    lock: 0x55a40aed2080:

    lock: 0x55a40aed2380:

    lock: 0x55a40aed2980:

    lock: 0x55a40aed2a80:

    lock: 0x55a40aed2b80:

    lock: 0x55a40aed1f80:

    lock: 0x55a40aed1c80:

    lock: 0x55a40aed1e80:

    lock: 0x55a40aed1b80:

    lock: 0x55a40aed1d80:

    lock: 0x55a40aed2680:

    lock: 0x55a40aed2780:

    lock: 0x55a40aed2880:

    lock: 0x55a40aed4780:

    lock: 0x55a40aed4a80:

    lock: 0x55a40aed4980:

    lock: 0x55a40aed4880:

    lock: 0x55a40aed4380:

    lock: 0x55a40aed4280:

    lock: 0x55a40aed4180:

    lock: 0x55a40aed4080:

    lock: 0x55a40aed3f80:

    lock: 0x55a40aed4580:

    lock: 0x55a40aed3e80:

    lock: 0x55a40aed3d80:

    lock: 0x55a40aed4680:

    lock: 0x55a40aed3b80:

    lock: 0x55a40aed3680:

    lock: 0x55a40aed3780:

    lock: 0x55a40aed3580:

    lock: 0x55a40aed3480:

    lock: 0x55a40aed3080:

    lock: 0x55a40aed3380:

    lock: 0x55a40aed3280:

    lock: 0x55a40aed2f80:

    lock: 0x55a40aed3180:

    lock: 0x55a40aed2c80:

    lock: 0x55a40aed2d80:

    lock: 0x55a40aed2e80:

    lock: 0x55a40aed4480:


    Key caches:
    default
    Buffer_size:     134217728
    Block_size:           1024
    Division_limit:        100
    Age_threshold:         300
    Partitions:              0
    blocks used:             0
    not flushed:             0
    w_requests:              0
    writes:                  0
    r_requests:              0
    reads:                   0


    handler status:
    read_key:         6553
    read_next:        1941
    read_rnd           262
    read_first:          1
    write:               1
    delete               0
    update:              0

    Table status:
    Opened tables:         67
    Open tables:          130
    Open files:            24
    Open streams:           0

    Alarm status:
    Active alarms:   0
    Max used alarms: 0
    Next alarm time: 0

    Memory status:
    Non-mmapped space allocated from system: 0
    Number of free chunks:                   1
    Number of fastbin blocks:                0
    Number of mmapped regions:               0
    Space in mmapped regions:                0
    Maximum total allocated space:           0
    Space available in freed fastbin blocks: 0
    Total allocated space:                   0
    Total free space:                        0
    Top-most, releasable space:              0
    Estimated memory (with thread stack):    0



    Events status:
    LLA = Last Locked At  LUA = Last Unlocked At
    WOC = Waiting On Condition  DL = Data Locked

    Event scheduler status:
    State      : INITIALIZED
    Thread id  : 0
    LLA        : n/a:0
    LUA        : n/a:0
    WOC        : NO
    Workers    : 0
    Executed   : 0
    Data locked: NO

    Event queue status:
    Element count   : 0
    Data locked     : NO
    Attempting lock : NO
    LLA             : init_queue:140
    LUA             : init_queue:150
    WOC             : NO
    Next activation : never
    160109  8:41:12 [Warning] 'proxies_priv' entry '@% root@609b1a101af3' ignored in
     --skip-name-resolve mode.
    FATA[0026] Error response from daemon: no such id: dfc9657da7365d03db58f6de73c95
    43a65100162332e585da9294472ac95082e

On the screen in Kitematic I have

![image](https://cloud.githubusercontent.com/assets/5702163/12215271/1e1d2ca4-b66a-11e5-92e3-f034f29cc922.png)

Note that I cannot click on the "Start" box any more. Doing docker ps shows the following.

    PS C:\users\akent> docker ps -a
    CONTAINER ID        IMAGE               COMMAND             CREATED
    STATUS              PORTS               NAMES
    9038c37214d0        gsd                 "/entrypoint.sh"    5 minutes ago
    Created                                 gsd

**Comment by FrenchBen** [comment]
Couple of things - When the container fails within kitematic, it also means that the listener may also be blocked. Essentially your actions via command line no longer propagate the update.

Secondly what image are you trying to use? What does your grunt/live-reload do? Does it affect the running containers?

Adding refresh when the 'Containers' title is clicked shouldn't be too difficult.

**Comment by FrenchBen** [comment]
Unless you see it being completely different, I'm going to close this issue in favor of https://github.com/docker/kitematic/issues/1331

**Comment by alankent** [comment]
You can choose - just reporting problems in case helpful. I think the sync problem may be the same. Do you think that would could bind errors report ports already in use? If it is because the container is actually running but kinematic is just out of sync, then happy for you to fix the sync problem then see if this problem goes away.

**Comment by FrenchBen** [comment]
when the container starts, it chooses a random port - If the port you're trying to assign is already in use, then it should show an error, preventing you from using it and also saving anything. 

This looks a bit more like a sync issue, although it's difficult to tell without replicating 100% - Feel free to create a step by step of how to reproduce with the matching images.

**Comment by alankent** [comment]
Hi, I generated a simpler repeatable test case. It is very easy to repeat on my laptop using the following.

Dockerfile:

    FROM php:7.0-apache
    EXPOSE 35729
    VOLUME /test

A test.bat file I run to redo it (this is on Windows):

    docker kill xxx
    docker rm xxx
    docker rmi xxx
    docker build -t xxx .
    docker run -p 8000:80 --name xxx xxx

1. I run the above commands, and the 'xxx' container appears in Kitematic.
2. I click on the volume and select 'enable volume' from the popup. The following error comes up.

![image](https://cloud.githubusercontent.com/assets/5702163/12273276/d2a9c0f4-b919-11e5-8464-504de5d0e21f.png)

At that stage I cannot click the "start" or "restart" buttons etc to restart the image from the UI. The buttons are greyed out.

Experimenting some more from command line:

    PS> docker start xxx
    Error response from daemon: Cannot start container xxx: failed to create endpoint xxx on network bridge: Bind for 0.0.0.0:8000 failed: port is already allocated
    
    FATA[0000] Error: failed to start one or more containers
    PS> docker ps -a
    CONTAINER ID        IMAGE               COMMAND                CREATED
       STATUS                        PORTS                NAMES
    43a92d78dc0b        xxx                 "apache2-foreground"   8 minutes ago
       Created                                            xxx
    PS> docker stop xxx
    xxx
    PS> docker stop xxx
    xxx
    PS> docker start xxx
    xxx
    PS> docker ps -a
    CONTAINER ID        IMAGE               COMMAND                CREATED
       STATUS                        PORTS                             NAMES
    43a92d78dc0b        xxx                 "apache2-foreground"   10 minutes ago
       Up About a minute             35729/tcp, 0.0.0.0:8000->80/tcp   xxx

    PS> docker kill xxx
    xxx
    PS> docker rm xxx
    xxx
    PS> docker rmi xxx
    Untagged: xxx:latest
    Deleted: e16afe4eda9fce6fc7b9f3541b7eb50f6a20c69be575f2113237981f7f30700c
    Deleted: 9e9a19bfa569c42caf1572cde495014ee746ce8d80ace68af963335beabd2f55
    PS>

During none of the above did the Kitematic UI update. I had to restart Kitematic to get it to work again.

I will let you choose if to repoen or same as other problem.  I think the caching problem is related, but I don't understand why restarting such a simple container to mount a volume causes the failure.

**Comment by FrenchBen** [comment]
Thanks for the steps to reproduce. 
Bad news:
* I was able to replicate exactly with the latest toolbox. 

Good news:
* This is fixed in the latest kitematic (master branch)

If you don't mind doing some QA, could you try this Kitematic windows version?
http://expirebox.com/download/0e55403fffd8892b222971beb018313f.html

PS: The above download link will expire in 48hrs

**Comment by alankent** [comment]
I unzipped and ran the binary - things were better, but a little strange. To be sure, I rebooted my laptop and started fresh. This time it all looked good and sane. I can mount a directory and it works. Thanks!

**Comment by FrenchBen** [comment]
Glad to here it all worked out - Thanks for the QA. A new toolbox should be released soon.
